components/script/dom/customelementregistry.rs

```diff
@@ -0,0 +1,303 @@
+            let cx = promise.global().get_cx();
+            // 16.2
+            promise.resolve_native(cx, &UndefinedValue());
+        }
+        Ok(())
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-customelementregistry-get
+    #[allow(unsafe_code)]
+    unsafe fn Get(&self, cx: *mut JSContext, name: DOMString) -> JSVal {
+        match self.definitions.borrow_mut().get(&name) {
+            Some(definition) => {
+                rooted!(in(cx) let mut constructor = UndefinedValue());
+                definition.constructor.to_jsval(cx, constructor.handle_mut());
+                constructor.get()
+            },
+            None => UndefinedValue(),
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-customelementregistry-whendefined
+    #[allow(unrooted_must_root)]
+    fn WhenDefined(&self, name: DOMString) -> Fallible<Rc<Promise>> {
+        let global_scope = self.window.upcast::<GlobalScope>();
+
+        // Step 1
+        if !is_valid_custom_element_name(&name) {
+            return Err(Error::Syntax);
+        }
+
+        // Step 2
+        if self.definitions.borrow().contains_key(&name) {
+            let promise = Promise::new(global_scope);
+            promise.resolve_native(global_scope.get_cx(), &UndefinedValue());
+            return Ok(promise)
+        }
+
+        // Step 3
+        let mut map = self.when_defined.borrow_mut();
+
+        // Steps 4, 5
+        let promise = match map.get(&name).cloned() {
+            Some(promise) => promise,
+            None => {
+                let promise = Promise::new(global_scope);
+                map.insert(name, promise.clone());
+                promise
+            }
+        };
+
+        // Step 6
+        Ok(promise)
+    }
+}
+
+#[derive(HeapSizeOf, JSTraceable)]
+struct CustomElementDefinition {
+    name: DOMString,
+
+    local_name: DOMString,
+
+    #[ignore_heap_size_of = "Rc"]
+    constructor: Rc<Function>,
+}
+
+impl CustomElementDefinition {
+    fn new(name: DOMString, local_name: DOMString, constructor: Rc<Function>) -> CustomElementDefinition {
+        CustomElementDefinition {
+            name: name,
+            local_name: local_name,
+            constructor: constructor,
+        }
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/#valid-custom-element-name
+fn is_valid_custom_element_name(name: &str) -> bool {
+    // Custom elment names must match:
+    // PotentialCustomElementName ::= [a-z] (PCENChar)* '-' (PCENChar)*
+
+    if let Some(c) = name.chars().nth(0) {
+        if c < 'a' || c > 'z' {
+            return false;
+        }
+    } else {
+        return false;
+    }
+
+    let mut has_dash = false;
+
+    for c in name.chars() {
+        if c == '-' {
+            has_dash = true;
+        }
+
+        // Check if this character is not a PCENChar
+        // https://html.spec.whatwg.org/multipage/#prod-pcenchar
+        if c != '-' && c != '.' && c != '_' && c != '\u{B7}' &&
```

Comment:
Please make a `is_potential_custom_element_char` function and use that here, all the reversed conditions don't help readability.